extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/shuffle/output/ChannelStageOutputReader.java

```diff
@@ -0,0 +1,243 @@
+                cursor,
+                offset
+            );
+          }
+        } else if (channel.isFinished()) {
+          try {
+            writer.close();
+          }
+          catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+
+          didCloseWriter = true;
+          continue;
+        } else if (channel.canRead()) {
+          try {
+            writer.writeFrame(channel.read(), FrameFileWriter.NO_PARTITION);
+          }
+          catch (Exception e) {
+            try {
+              writer.abort();
+            }
+            catch (IOException e2) {
+              e.addSuppressed(e2);
+            }
+
+            throw new RuntimeException(e);
+          }
+        } else {
+          return FutureUtils.transformAsync(channel.readabilityFuture(), ignored -> readRemotelyFrom(offset));
+        }
+      }
+
+      // Remove first chunk if it is no longer needed. (offset is entirely past it.)
+      final byte[] chunk = chunks.peek();
+      final long amountToAdvance = Math.min(offset - cursor, chunk.length - positionWithinFirstChunk);
+      cursor += amountToAdvance;
+      positionWithinFirstChunk += Ints.checkedCast(amountToAdvance);
+
+      if (positionWithinFirstChunk == chunk.length) {
+        chunks.poll();
+        positionWithinFirstChunk = 0;
+      }
+    }
+
+    if (chunks.isEmpty() || offset != cursor) {
+      throw DruidException.defensive(
+          "Expected cursor[%,d] to be caught up to offset[%,d] by this point, and to have nonzero chunks",
+          cursor,
+          offset
+      );
+    }
+
+    return Futures.immediateFuture(new ByteChunksInputStream(ImmutableList.copyOf(chunks), positionWithinFirstChunk));
+  }
+
+  @Override
+  public synchronized ReadableFrameChannel readLocally()
+  {
+    if (state == State.INIT) {
+      state = State.LOCAL;
+      return channel;
+    } else if (state == State.REMOTE) {
+      throw new ISE("Cannot read both remotely and locally");
+    } else if (state == State.LOCAL) {
+      throw new ISE("Cannot read channel multiple times");
+    } else {
+      assert state == State.CLOSED;
+      throw new ISE("Closed");
+    }
+  }
+
+  @Override
+  public synchronized void close()
+  {
+    // Call channel.close() unless readLocally() has been called. In that case, we expect the caller to close it.
+    if (state != State.LOCAL) {
+      channel.close();
+    }
+  }
+
+  /**
+   * Input stream that can have bytes appended to it, and that can have bytes acknowledged.
+   */
+  private class ChunkAcceptor implements WritableByteChannel
+  {
+    private boolean open = true;
+
+    @Override
+    public int write(final ByteBuffer src) throws IOException
+    {
+      if (!open) {
+        throw new IOException("Closed");
+      }
+
+      final int len = src.remaining();
+      if (len > 0) {
+        final byte[] bytes = new byte[len];
+        src.get(bytes);
+        chunks.add(bytes);
```

Comment:
This makes additional copies of the bytes. I am curious if 
a. there's a way to reduce this additional copy. 
b. Does having this copy add any meaningful overhead
c. If there's an overhead, was it present in the original code, or does the refactor introduce it.

Reply:
my thoughts:

a) I think the copy can be eliminated if we alter `FrameFileWriter` somehow. It accepts a `WritableByteChannel` that it writes the frame file to, and the `WritableByteChannel` method `int write(ByteBuffer src)` requires copying the bytes from `src`. (The ByteBuffer object cannot be retained, since the caller might reuse it.) So, we'd need to change `FrameFileWriter` to have some option other than `WritableByteChannel`.

b) I haven't done extensive profiling, so I'm not sure. I was planning to do more extensive profiling and look into further perf enhancements after this code lands in master.

c) If there's any overhead, it wasn't present in the original code, because this entire code path is new. This code path is only used for in-memory shuffles. The original code always read from on-disk frame files, & did not generate them on-the-fly.